website/versioned_docs/version-v3.9.x/expansion.md

```diff
@@ -0,0 +1,218 @@
+for the `Pod`, runs validation on it, and aggregates the mock resource's
+violations onto the parent resource (the `Deployment` in this example).
+
+To use this functionality, users must first specify which resources should be "
+expanded" by creating 1 or more `TemplateExpansion` custom resources. Then, in
+order for Gatekeeper to accurately create these "mock resources" in a way that
+mirrors the real Kubernetes controllers, users can define any number of
+[Mutations](https://open-policy-agent.github.io/gatekeeper/website/docs/mutation)
+on the expanded resource to manipulate them into the desired form.
+
+Any resources configured for expansion will be expanded for both validation and
+[Audit](https://open-policy-agent.github.io/gatekeeper/website/docs/audit). This
+feature will only be enabled if a user creates a `TemplateExpansion`
+that targets some resource that exists on the cluster.
+
+Note that the accuracy of enforcement depends on how well the mock resource
+resembles the real thing. Mutations can help with this, but 100% accuracy is
+impossible because not all fields can be predicted. For instance, deployments
+create pods with random names. Inaccurate mocks may lead to over or under
+enforcement. In the case of under enforcement, the resultant pod should still be
+rejected. Finally, non-state-based policies (those that rely on transient
+metadata such as requesting user or time of creation) cannot be enforced
+accurately. This is because such metadata would necessarily be different when
+creating the resultant resource. For example, a deployment is created using the
+requesting user's account, but the pod creation request comes from the service
+account of the deployment controller.
+
+If, for any reason, you want to exempt expanded resources from a specific
+constraint, look at the [Match Source](#match-source) section below.
+
+## Configuring Expansion
+
+Expansion behavior is configured through an `ExpansionTemplate` CR. Optionally,
+users can create `Mutation` CRs to customize how resources are expanded.
+
+Users can test their expansion configuration using the `gator expand` CLI (TODO
+link).
+
+#### ExpansionTemplate
+
+The `ExpansionTemplate` CR specifies:
+
+- Which resource(s) should be expanded, specified by their GVK
+- The GVK of the resultant resources
+- Which field to use as the "source" for the resultant resource. The source is a
+  field on the parent resource which will be used as the base for expanding it
+  before any mutators are applied. For example, in a case where a `Deployment`
+  expands into a `Pods`, `spec.template` would typically be the source.
+
+Here is an example of a `ExpansionTemplate` that specifies that `Deployment`
+and `ReplicaSet` should be expanded into a `Pod`.
+
+```
+apiVersion: expansion.gatekeeper.sh/v1alpha1
+kind: ExpansionTemplate
+metadata:
+  name: expand-deployments
+spec:
+  applyTo:
+    - groups: ["apps"]
+      kinds: ["Deployment", "ReplicaSet"]
```

Comment:
I know this is an example but would we want to add other top level kinds that create pods?

Reply:
Here are the list of resources that use Pod templates:

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.24/#podtemplatespec-v1-core

Particularly interesting is [CronJob](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.24/#cronjob-v1-batch), which has a different path for `templateSource` (all the rest are `spec.template`)

Reply:
Thanks for the reference Max. Added all resources using Pod templates.